src/envs/playroom.py

```diff
@@ -0,0 +1,845 @@
+                or (0.0 < x < 30.0 and 0.0 < y < 30.0) \
+                or (110.0 < x < 140.0 and 0.0 < y < 30.0):
+                if self._robot_can_move(state, action):
+                    transition_fn = self._transition_move
+            else:  # Invalid
+                return state
+        # In hallway not next to a door
+        elif (30.0 < prev_x < 110.0) and (10.0 < prev_y < 20.0):
+            # Move elsewhere in hallway or to boring room or to playroom
+            if (30.0 < x < 110.0 and 10.0 < y < 20.0) \
+                or (0.0 < x < 30.0 and 0.0 < y < 30.0) \
+                or (110.0 < x < 140.0 and 0.0 < y < 30.0):
+                if self._robot_can_move(state, action):
+                    transition_fn = self._transition_move
+            else:  # Invalid
+                return state
+        # In playroom next to dial
+        elif (110.0 < prev_x < 140.0) and (0.0 < prev_y < 30.0) and \
+                self._NextToDial_holds(state, (self._robot, self._dial)):
+            dial_x = state.get(self._dial, "pose_x")
+            dial_y = state.get(self._dial, "pose_y")
+            # Toggle dial
+            if (dial_x-cls.dial_button_tol < x < dial_x+cls.dial_button_tol) \
+                and (dial_y-cls.dial_button_tol < y
+                     < dial_y+cls.dial_button_tol) \
+                and self._robot_is_facing_dial(state, action):
+                next_state = state.copy()
+                if state.get(self._dial, "level") < 0.5:
+                    next_state.set(self._dial, "level", 1.0)
+                else:
+                    next_state.set(self._dial, "level", 0.0)
+                return next_state
+            # Move elsewhere in playroom
+            if (110.0 < x < 140.0) and (0.0 < y < 30.0):
+                transition_fn = self._transition_move
+            # Move to hallway or to boring room
+            elif (30.0 < x < 110.0 and 10.0 < y < 20.0) \
+                    or (0.0 < x < 30.0 and 0.0 < y < 30.0):
+                if self._robot_can_move(state, action):
+                    transition_fn = self._transition_move
+            else:  # Invalid
+                return state
+        # In playroom next to door
+        elif (110.0 < prev_x < 140.0) and (0.0 < prev_y < 30.0) and \
+            self._NextToDoor_holds(state, (self._robot, self._door6)):
+            door_x = state.get(self._door6, "pose_x")
+            door_y = state.get(self._door6, "pose_y")
+            # Open/close door
+            if (door_x-cls.door_tol < x < door_x + cls.door_tol) and \
+                (door_y-cls.door_tol < y < door_y+cls.door_tol) \
+                and (cls.door_button_z-cls.door_tol < z
+                     < cls.door_button_z+cls.door_tol) \
+                and (rotation >= 1.5 or rotation <= -1.5):
+                transition_fn = self._transition_door
+            # Move elsewhere in playroom
+            elif (110.0 < x < 140.0) and (0.0 < y < 30.0):
```

Comment:
Why are we repeating these same kinds of checks so many times? x and y are coming out of the action, so at the start of this function we could just create all the boolean variables we need, e.g. `is_in_playroom = (110.0 < x < 140.0) and (0.0 < y < 30.0)`